server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/BinaryValuesSource.java

```diff
@@ -0,0 +1,137 @@
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.CheckedFunction;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+
+import java.io.IOException;
+
+/**
+ * A {@link SingleDimensionValuesSource} for binary source ({@link BytesRef}).
+ */
+class BinaryValuesSource extends SingleDimensionValuesSource<BytesRef> {
+    private final CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValuesFunc;
+    private final BytesRef[] values;
+    private BytesRef currentValue;
+
+    BinaryValuesSource(MappedFieldType fieldType, CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValuesFunc,
+                       int size, int reverseMul) {
+        super(fieldType, size, reverseMul);
+        this.docValuesFunc = docValuesFunc;
+        this.values = new BytesRef[size];
+    }
+
+    @Override
+    String type() {
+        return "binary";
+    }
+
+    @Override
+    public void copyCurrent(int slot) {
+        values[slot] = BytesRef.deepCopyOf(currentValue);
+    }
+
+    @Override
+    public int compare(int from, int to) {
+        return compareValues(values[from], values[to]);
+    }
+
+    @Override
+    int compareCurrent(int slot) {
+        return compareValues(currentValue, values[slot]);
+    }
+
+    @Override
+    int compareCurrentWithAfter() {
+        return compareValues(currentValue, afterValue);
+    }
+
+    int compareValues(BytesRef v1, BytesRef v2) {
+        return v1.compareTo(v2) * reverseMul;
+    }
+
+    @Override
+    void setAfter(Comparable<?> value) {
+        if (value.getClass() == BytesRef.class) {
+            afterValue = (BytesRef) value;
+        } else if (value.getClass() == String.class) {
+            afterValue = new BytesRef((String) value);
+        } else {
+            throw new IllegalArgumentException("invalid value, expected string, got " + value.getClass().getSimpleName());
+        }
+    }
+
+    @Override
+    BytesRef toComparable(int slot) {
+        return values[slot];
+    }
+
+    @Override
+    LeafBucketCollector getLeafCollector(LeafReaderContext context, LeafBucketCollector next) throws IOException {
+        final SortedBinaryDocValues dvs = docValuesFunc.apply(context);
+        return new LeafBucketCollector() {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                if (dvs.advanceExact(doc)) {
+                    int num = dvs.docValueCount();
+                    for (int i = 0; i < num; i++) {
+                        currentValue = dvs.nextValue();
```

Comment:
this means currentValue will always be the higher value in case of a multi-valued field, is that ok?

Reply:
currentValue is only valid for the current composite bucket, `next.collect()` below will fill the other sources's currentValue and the last collector in the chain will check if the final composite bucket should be added in the queue. We don't use currentValue outside of these recursive calls.

Reply:
I see, thanks.